File: apps/web/src/types/conversation.ts
/**
 * ConversationRelatedTypeDefinition
 */

// ===== MessageRole =====
export type MessageRole = "user" | "assistant" | "system";

// ===== MessageAttachmentType =====
export interface MessageAttachment {
 id: string;
 type: "image" | "file" | "code";
 name: string;
 url: string;
 size: number;
 mimeType: string;
}

// ===== Message =====
export interface Message {
 id: string;
 conversationId: string;
 role: MessageRole;
 content: string;
 model?: string;
 tokenUsage: number;
 promptTokens: number;
 completionTokens: number;
 attachments?: MessageAttachment[];
 metadata?: Record<string, unknown>;
 parentId?: string;
 // UserFeedback
 liked?: boolean;
 disliked?: boolean;
 bookmarked?: boolean;
 createdAt: string;
 updatedAt: string;
}

// ===== AI ParameterSettings =====
export interface AIParameters {
 temperature?: number; // 0.0-2.0, Default 1.0
 maxTokens?: number; // MaximumGenerate token count
 topP?: number; // 0.0-1.0
 topK?: number; // Top-K Sampling
 frequencyPenalty?: number; // -2.0-2.0
 presencePenalty?: number; // -2.0-2.0
}

// ===== Conversation =====
export interface Conversation extends AIParameters {
 id: string;
 userId: string;
 workspaceId?: string;
 title: string;
 preview: string;
 model: string;
 systemPrompt?: string;
 starred: boolean;
 pinned: boolean;
 archived: boolean;
 messageCount: number;
 tokenUsage: number;
 folderId?: string;
 metadata?: Record<string, unknown>;
 tags: string[];
 createdAt: string;
 updatedAt: string;
 messages?: Message[];
 folder?: ConversationFolder;
}

// ===== ConversationFolder =====
export interface ConversationFolder {
 id: string;
 userId: string;
 name: string;
 icon: string;
 color: string;
 parentId?: string;
 sortOrder: number;
 conversationCount: number;
 createdAt: string;
 updatedAt: string;
}

// ===== RequestType =====

export interface CreateConversationRequest extends AIParameters {
 workspaceId: string;
 title: string;
 model?: string;
 systemPrompt?: string;
 folderId?: string;
 tags?: string[];
}

export interface UpdateConversationRequest extends AIParameters {
 title?: string;
 model?: string;
 systemPrompt?: string;
 folderId?: string;
}

export interface ListConversationsParams {
 workspaceId?: string;
 folderId?: string;
 starred?: boolean;
 pinned?: boolean;
 archived?: boolean;
 search?: string;
 page?: number;
 pageSize?: number;
 orderBy?: string;
}

export interface AddMessageRequest {
 role: MessageRole;
 content: string;
 model?: string;
 tokenUsage?: number;
 promptTokens?: number;
 completionTokens?: number;
 parentId?: string; // Reply/use'sMessage ID
}

export interface ListMessagesParams {
 page?: number;
 pageSize?: number;
 beforeId?: string;
 afterId?: string;
}

export interface BatchOperationRequest {
 ids: string[];
}

export interface BatchStarRequest {
 ids: string[];
 starred: boolean;
}

export interface BatchArchiveRequest {
 ids: string[];
 archived: boolean;
}

export interface BatchMoveRequest {
 ids: string[];
 folderId?: string;
}

export interface CreateConversationFolderRequest {
 name: string;
 icon?: string;
 color?: string;
 parentId?: string;
}

export interface UpdateConversationFolderRequest {
 name?: string;
 icon?: string;
 color?: string;
 sortOrder?: number;
}

// ===== ResponseType =====

export interface ConversationListResponse {
 conversations: Conversation[];
 total: number;
 page: number;
 pageSize: number;
}

export interface MessageListResponse {
 messages: Message[];
 total: number;
 page: number;
 pageSize: number;
}

export interface ConversationFolderListResponse {
 folders: ConversationFolder[];
 total: number;
}

export interface BatchOperationResponse {
 success: boolean;
 count: number;
 message: string;
}

// ===== AI ModelOption =====
export interface AIModel {
 id: string;
 name: string;
 provider: string;
 description?: string;
 maxTokens: number;
 contextLength: number;
 pricing?: {
 input: number;
 output: number;
 };
}

export const AI_MODELS: AIModel[] = [
 {
 id: "gpt-4",
 name: "GPT-4",
 provider: "OpenAI",
 description: "Bestlarge'sGPTModel, SuitableComplexTask",
 maxTokens: 8192,
 contextLength: 8192,
 },
 {
 id: "gpt-4-turbo",
 name: "GPT-4 Turbo",
 provider: "OpenAI",
 description: "moreQuick'sGPT-4Version",
 maxTokens: 128000,
 contextLength: 128000,
 },
 {
 id: "gpt-3.5-turbo",
 name: "GPT-3.5 Turbo",
 provider: "OpenAI",
 description: "compareSelect",
 maxTokens: 16385,
 contextLength: 16385,
 },
 {
 id: "claude-3-opus",
 name: "Claude 3 Opus",
 provider: "Anthropic",
 description: "Anthropic BestModel",
 maxTokens: 200000,
 contextLength: 200000,
 },
 {
 id: "claude-3-sonnet",
 name: "Claude 3 Sonnet",
 provider: "Anthropic",
 description: "BalancecanandSpeed",
 maxTokens: 200000,
 contextLength: 200000,
 },
 {
 id: "claude-3-haiku",
 name: "Claude 3 Haiku",
 provider: "Anthropic",
 description: "QuickResponse",
 maxTokens: 200000,
 contextLength: 200000,
 },
];

// ===== ModelIconMapping =====
export const MODEL_ICONS: Record<string, string> = {
 "gpt-4": "✨",
 "gpt-4-turbo": "⚡",
 "gpt-3.5-turbo": "💫",
 "claude-3-opus": "🎭",
 "claude-3-sonnet": "🎵",
 "claude-3-haiku": "🌸",
};

// ===== Toolcount =====

export function getModelDisplayName(modelId: string): string {
 const model = AI_MODELS.find((m) => m.id === modelId);
 return model?.name || modelId;
}

export function getModelIcon(modelId: string): string {
 return MODEL_ICONS[modelId] || "🤖";
}

export function formatMessagePreview(content: string, maxLength = 200): string {
 if (content.length <= maxLength) return content;
 return content.slice(0, maxLength) + "...";
}

export function formatRelativeTime(dateString: string): string {
 const date = new Date(dateString);
 const now = new Date();
 const diffMs = now.getTime() - date.getTime();
 const diffSec = Math.floor(diffMs / 1000);
 const diffMin = Math.floor(diffSec / 60);
 const diffHour = Math.floor(diffMin / 60);
 const diffDay = Math.floor(diffHour / 24);

 if (diffSec < 60) return "Just now";
 if (diffMin < 60) return `${diffMin}minbefore`;
 if (diffHour < 24) return `${diffHour}hbefore`;
 if (diffDay < 7) return `${diffDay}daysbefore`;
 
 return date.toLocaleDateString("zh-CN");
}

File: apps/server/internal/domain/entity/conversation.go
package entity

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Conversation 对话实体
type Conversation struct {
	ID          uuid.UUID `gorm:"type:char(36);primaryKey" json:"id"`
	UserID      uuid.UUID `gorm:"type:char(36);not null;index" json:"user_id"`
	WorkspaceID uuid.UUID `gorm:"type:char(36);not null;index" json:"workspace_id"`

	// 基础信息
	Title   string `gorm:"size:500;not null" json:"title"`
	Preview string `gorm:"size:500" json:"preview"` // 最后一条消息预览

	// AI 模型配置
	Model        string  `gorm:"size:50;default:'gpt-4'" json:"model"`
	SystemPrompt *string `gorm:"type:text" json:"system_prompt"`

	// AI 参数设置
	Temperature      *float64 `gorm:"type:decimal(3,2)" json:"temperature"`       // 0.0-2.0，默认 1.0
	MaxTokens        *int     `json:"max_tokens"`                                 // 最大生成 token 数
	TopP             *float64 `gorm:"type:decimal(3,2)" json:"top_p"`             // 0.0-1.0
	TopK             *int     `json:"top_k"`                                      // Top-K 采样
	FrequencyPenalty *float64 `gorm:"type:decimal(3,2)" json:"frequency_penalty"` // -2.0-2.0
	PresencePenalty  *float64 `gorm:"type:decimal(3,2)" json:"presence_penalty"`  // -2.0-2.0

	// 状态标记
	Starred  bool `gorm:"default:false;index" json:"starred"`
	Pinned   bool `gorm:"default:false;index" json:"pinned"`
	Archived bool `gorm:"default:false;index" json:"archived"`

	// 统计信息
	MessageCount int `gorm:"default:0" json:"message_count"`
	TokenUsage   int `gorm:"default:0" json:"token_usage"`

	// 文件夹
	FolderID *uuid.UUID `gorm:"type:char(36);index" json:"folder_id"`

	// 元数据 (JSON)
	Metadata JSON `gorm:"type:json" json:"metadata"`

	// 时间戳
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	User      *User               `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Workspace *Workspace          `gorm:"foreignKey:WorkspaceID" json:"workspace,omitempty"`
	Folder    *ConversationFolder `gorm:"foreignKey:FolderID" json:"folder,omitempty"`
	Messages  []Message           `gorm:"foreignKey:ConversationID" json:"messages,omitempty"`
	Tags      []ConversationTag   `gorm:"foreignKey:ConversationID" json:"tags,omitempty"`
}

// TableName 表名
func (Conversation) TableName() string {
	return "what_reverse_conversations"
}

// BeforeCreate 创建前钩子
func (c *Conversation) BeforeCreate(tx *gorm.DB) error {
	if c.ID == uuid.Nil {
		c.ID = uuid.New()
	}
	if c.Model == "" {
		c.Model = "gpt-4"
	}
	return nil
}

// ConversationFolder 对话文件夹实体
type ConversationFolder struct {
	ID                uuid.UUID      `gorm:"type:char(36);primaryKey" json:"id"`
	UserID            uuid.UUID      `gorm:"type:char(36);not null;index" json:"user_id"`
	Name              string         `gorm:"size:100;not null" json:"name"`
	Icon              string         `gorm:"size:50;default:'📁'" json:"icon"`
	Color             string         `gorm:"size:20;default:'#3ECF8E'" json:"color"`
	ParentID          *uuid.UUID     `gorm:"type:char(36);index" json:"parent_id"`
	SortOrder         int            `gorm:"default:0" json:"sort_order"`
	ConversationCount int            `gorm:"-" json:"conversation_count"` // 计算字段
	CreatedAt         time.Time      `json:"created_at"`
	UpdatedAt         time.Time      `json:"updated_at"`
	DeletedAt         gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	User          *User                `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Parent        *ConversationFolder  `gorm:"foreignKey:ParentID" json:"parent,omitempty"`
	Children      []ConversationFolder `gorm:"foreignKey:ParentID" json:"children,omitempty"`
	Conversations []Conversation       `gorm:"foreignKey:FolderID" json:"conversations,omitempty"`
}

// TableName 表名
func (ConversationFolder) TableName() string {
	return "what_reverse_conversation_folders"
}

// BeforeCreate 创建前钩子
func (f *ConversationFolder) BeforeCreate(tx *gorm.DB) error {
	if f.ID == uuid.Nil {
		f.ID = uuid.New()
	}
	if f.Icon == "" {
		f.Icon = "📁"
	}
	if f.Color == "" {
		f.Color = "#3ECF8E"
	}
	return nil
}

// ConversationTag 对话标签关联实体
type ConversationTag struct {
	ID             uuid.UUID `gorm:"type:char(36);primaryKey" json:"id"`
	ConversationID uuid.UUID `gorm:"type:char(36);not null;index" json:"conversation_id"`
	TagName        string    `gorm:"size:50;not null;index" json:"tag_name"`
	CreatedAt      time.Time `json:"created_at"`
}

// TableName 表名
func (ConversationTag) TableName() string {
	return "what_reverse_conversation_tags"
}

// BeforeCreate 创建前钩子
func (t *ConversationTag) BeforeCreate(tx *gorm.DB) error {
	if t.ID == uuid.Nil {
		t.ID = uuid.New()
	}
	return nil
}
